File: app/src/main/java/closer/vlllage/com/closer/handler/helpers/InstallShortcutHandler.java
package closer.vlllage.com.closer.handler.helpers;

import android.graphics.Bitmap;
import android.graphics.Color;
import android.support.v4.content.pm.ShortcutInfoCompat;
import android.support.v4.content.pm.ShortcutManagerCompat;
import android.support.v4.graphics.drawable.IconCompat;

import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.handler.group.GroupActivityTransitionHandler;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.store.models.Group;

public class InstallShortcutHandler extends PoolMember {
    public void installShortcut(Group group) {
        if (ShortcutManagerCompat.isRequestPinShortcutSupported($(ApplicationHandler.class).getApp())) {
            Bitmap icon = $(ShortcutIconGenerator.class).generate(
                    $(Val.class).of(group.getName(), $(ResourcesHandler.class).getResources().getString(R.string.app_name)),
                    128,
                    Color.WHITE,
                    $(GroupColorHandler.class).getColor(group),
                    $(GroupColorHandler.class).getLightColor(group)
            );
            ShortcutInfoCompat shortcutInfo = new ShortcutInfoCompat.Builder($(ApplicationHandler.class).getApp(), "group-" + group.getId())
                    .setIntent($(GroupActivityTransitionHandler.class).getIntent(group.getId(), false))
                    .setShortLabel($(Val.class).of(group.getName(), $(ResourcesHandler.class).getResources().getString(R.string.app_name)))
                    .setIcon(IconCompat.createWithBitmap(icon))
                    .build();
            ShortcutManagerCompat.requestPinShortcut($(ApplicationHandler.class).getApp(), shortcutInfo, null);
            icon.recycle();
        }
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/helpers/InstallShortcutHandler.kt
package closer.vlllage.com.closer.handler.helpers

import android.graphics.Color
import androidx.core.content.pm.ShortcutInfoCompat
import androidx.core.content.pm.ShortcutManagerCompat
import androidx.core.graphics.drawable.IconCompat
import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.handler.group.GroupActivityTransitionHandler
import closer.vlllage.com.closer.store.models.Group
import com.queatz.on.On
import io.reactivex.android.schedulers.AndroidSchedulers

class InstallShortcutHandler constructor(private val on: On) {
    fun installShortcut(group: Group) {
        if (ShortcutManagerCompat.isRequestPinShortcutSupported(on<ApplicationHandler>().app)) {
            on<GroupNameHelper>().getName(group)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ name ->
                on<DefaultInput>().show(R.string.icon_text, prefill = name.take(4), maxLength = 4) { iconText ->
                    val icon = on<ShortcutIconGenerator>().generate(
                        iconText,
                        128f,
                        Color.WHITE,
                        on<GroupColorHandler>().getColor(group),
                        on<GroupColorHandler>().getLightColor(group)
                    )
                    val shortcutInfo = ShortcutInfoCompat.Builder(on<ApplicationHandler>().app, "group-" + group.id!!)
                        .setIntent(on<GroupActivityTransitionHandler>().getIntent(group.id!!, false, isPhone = group.hasPhone()))
                        .setShortLabel(name)
                        .setIcon(IconCompat.createWithBitmap(icon))
                        .build()
                    ShortcutManagerCompat.requestPinShortcut(on<ApplicationHandler>().app, shortcutInfo, null)
                    icon.recycle()
                }
            }, {
                on<DefaultAlerts>().thatDidntWork()
            }).also {
                on<DisposableHandler>().add(it)
            }
        }
    }
}
